### ISSUE TITLE
PDP-15 I/O Device Expansion

### ISSUE THREAD

**Issue by cmgauger** [simh/simh#96]
I've recently begun screwing around with the PDP-15 simulator and the available operating systems again, and I've been thinking about how to go about implementing some of the peripherals the PDP-15 is missing.

First, let's go over the "missing" peripherals for the PDP-15:
1. CR15 or CR03B card reader
2. VT15 graphics display
3. VW01 writing tablet
4. UNICHANNEL-15 + associated peripherals

I personally have don't have a use for or, find I'm very interested in, the VT15 or VW01, so I'm going to ignore those (however, I do welcome individuals who are interested in them to contribute ideas).

So let's address the CR15 and CR03B readers first since they would probably be the simpler of the missing peripherals to implement.

I cannot find documentation on the CR15 reader and control, so I don't know how different it is in comparison to the CR03B; neither have I found documentation on the CR03B reader and control specifically, however, extrapolating on information present in the PDP-9 User Handbook, the CR03B reader control is probably of the same functionality as the control of the CR01B and CR02B, the difference between the three models of CR0xB reader being speed of operation (CR01B at 100 cards per minute, CR02B at 200 cards per minute, and I'm going to assume the CR03B was 300 cards per minute).

From the information in the PDP-9 handbook -- a description of the IOT instructions responsible for card reader control, and the general operation of the reader control and readers themselves, would it be possible to reconstruct the CR0xB reader of the PDP-9 and PDP-15?

As I mentioned earlier I am unsure of how the CR15 card reader and control match up with that of the CR0xB reader and control.

Now, I'm going to address the elephant in the room, namely the UNICHANNEL-15 option and its associated peripherals. I have asked on the mailing list with regards to SIMH and multiple processor systems before, including the UC15. While I'm not nearly competent enough to actually implement a device or anything approching multiple processor support, I have been thinking of ways to go about getting the PDP-15 it's quite popular UC15 attachment.

The UC15 at it's most basic and stripped down form, is a PDP-11/05, with 4KW or 8KW of local/private memory, and the remaining 24KW or 28KW of memory shared with the PDP-15. The PDP-11 is also equipped with two DR11C interfaces, and the PDP-15 with a DR15 interface. The PDP-11 runs a standalone software system known as PIREX loaded from paper tape by the PDP-15; through the DRxx devices the two computer systems may interrupt each other as well as present information and commands. (Data is transferred via the shared memory.) Depending on peripheral configuration, and operating system configuration, the PDP-11 might also be running the SPOL11 spooler program, and small MAC11 (the name of the PDP-11 assembler which is available in XVM/DOS and XVM/RSX) assembler programs written on the PDP-15.

The PDP-11 peripherals which were supported by the PDP-15 software are:
1. RK15 RK05 system
2. LP11 lineprinter
3. LS11 lineprinter
4. LV11 printer-plotter
5. CR11 card reader
6. XY11 plotter control

Of these, the LP11 and CR11 are already implemented in the SIMH PDP-11, and the RK15 is based upon the RK11, but more on that in a minute. I'm not entirely sure, but I do recall the LS11 lineprinter is unnecessary to implement due to being "the same" as the LP11 from a software perspective. The LV11 is an interesting device, however I do not see great use being made of it. Similarly so with the XY11, I see no great use being made of it, although a plotter "engine" if implemented would allow for implementing plotters in the '8 and other 18-bit systems. (Plotter "engine" as in something with could understand the movement commands and pen-up/pen-down commands and turn that into some sort of graphics file.)

The RK15 makes use of the RK11-E controller, which works much like the standard RK11 controller, except it provides the facilities for 18-bit DMA. Disk format is 12 sector packs with 256 18-bit words per sector.

Because of the complexity of the UC15 system, I have conceived of three possibilities for implementing its functionality in SIMH:
1. The Direct Approach -- Implementing multiprocessor support within SIMH, which will probably require a hefty rewrite of the simulator core, to allow the PDP-15 simulator to incorporate a stripped down PDP-11 processor and peripheral set. This approach would be the most friendly for the end users of the software; however this method would probably require the most work in the nitty gritty parts of SIMH, but would probably result in a great benefit for other simulators (by which I mean the VAX and PDP-11 simulators; if SIMH could simulate a KL10, then also for a theoretical KL10 simulator too).
2. The Indirect Approach -- Implementing within both the PDP-15 and PDP-11 simulators an interface to allow two separate simulator programs to interface. The PDP-11 would need a new CPU model option "UC-15", which would:
a. Disable all unsupported devices, while enabling the PDP-15 connection parts.
b. Enable 18-bit DMA on the part of the RK11.
c. "Start" the processor, and put the entire thing under the control of the PDP-15.
This would also probably require some large amount of work to implement. And while I personally think it would be easier to implement than native multiprocessor support within SIMH itself, it doesn't come with as many benefits.
3. PIREX/SPOL11 Emulation -- Eschewing implementing any from of intersimulator/interprocessor communication at all. Instead a new device driver is written which simulates the function of PIREX and the SPOL11 spooler; peripherals would have to be controlled through a la the RHA/RHB devices for the RP and TU controllers of the PDP-11. This method however doesn't allow for the running of PDP-11 software developed on the PDP-15. (Plus one would need to deal with the states of the PIREX and SPOL11 software.)

In conclusion and as a summary, at a minimum the CR0xB card readers for the PDP-9 and PDP-15 would be "relatively" easy to implement, however the CR15 card reader is lacking in information needed to implement it. Further the UNICHANNEL-15 option -- specifically the RK15 disk system -- would be an excellent thing to have added to the PDP-15, however the only options that I personally can see of implementing it are non-trivial.

Comments are welcome.

**Comment by archivisth** [comment]
I have finished an implementation of the UC15 using two separate processes - one a PDP15, one a PDP11 with an 18bit RK11 variant. Now I need to debug it and also document the rather complex operational procedures. 

Thanks to Mark P for providing the underlying shared memory libraries.

**Comment by archivisth** [comment]
This issue can be closed.
